builtin/providers/aws/resource_aws_instance.go

```diff
@@ -986,7 +997,15 @@ func buildAwsInstanceOpts(
 		}
 	}
 
-	if hasSubnet && associatePublicIPAddress {
+	if hasNetworkInterface {
+		ni := &ec2.InstanceNetworkInterfaceSpecification{
+			NetworkInterfaceId: aws.String(networkInterfaceID),
+			DeviceIndex:        aws.Int64(int64(0)),
+			SubnetId:           aws.String(subnetID),
```

Comment:
As far as the EC2 API is concerned, several of the parameters for an instance are effectively delegates for the parameters of the network interface that EC2 implicitly creates for the instance. In the case where existing network interfaces are explicitly supplied using `NetworkInterface.N` parameters, EC2 does not create an implicit network interfaces and these delegated parameters are invalid. Specifically, `PrivateIpAddress`, `SecurityGroup.N`, `SecurityGroupId.N`, and `SubnetId` are all mutually exclusive with `NetworkInterface.N`. Furthermore, the instance's `sourceDestCheck` attribute is inapplicable if network interfaces have been explicitly supplied to the instance.